src/colmap/estimators/generalized_pose.cc

```diff
@@ -0,0 +1,306 @@
+#include "colmap/estimators/pose.h"
+#include "colmap/geometry/rigid3.h"
+#include "colmap/math/matrix.h"
+#include "colmap/optim/ransac.h"
+#include "colmap/optim/support_measurement.h"
+#include "colmap/scene/camera.h"
+#include "colmap/sensor/models.h"
+#include "colmap/util/misc.h"
+
+#include <Eigen/Core>
+#include <ceres/ceres.h>
+
+namespace colmap {
+namespace {
+
+bool LowerVector3d(const Eigen::Vector3d& v1, const Eigen::Vector3d& v2) {
+  if (v1.x() < v2.x()) {
+    return true;
+  } else if (v1.x() == v2.x()) {
+    if (v1.y() < v2.y()) {
+      return true;
+    } else if (v1.y() == v2.y()) {
+      return v1.z() < v2.z();
+    } else {
+      return false;
+    }
+  } else {
+    return false;
+  }
+}
+
+}  // namespace
+
+bool EstimateGeneralizedAbsolutePose(
+    const RANSACOptions& options,
+    const std::vector<Eigen::Vector2d>& points2D,
+    const std::vector<Eigen::Vector3d>& points3D,
+    const std::vector<size_t>& camera_idxs,
+    const std::vector<Rigid3d>& cams_from_rig,
+    const std::vector<Camera>& cameras,
+    Rigid3d* rig_from_world,
+    size_t* num_inliers,
+    std::vector<char>* inlier_mask) {
+  CHECK_EQ(points2D.size(), points3D.size());
+  CHECK_EQ(points2D.size(), camera_idxs.size());
+  CHECK_EQ(cams_from_rig.size(), cameras.size());
+  CHECK_GE(*std::min_element(camera_idxs.begin(), camera_idxs.end()), 0);
+  CHECK_LT(*std::max_element(camera_idxs.begin(), camera_idxs.end()),
+           cameras.size());
+  options.Check();
+  if (points2D.size() == 0) {
+    return false;
+  }
+
+  const double max_error_px = options.max_error;
+  CHECK_GT(max_error_px, 0.0);
+
+  std::vector<GP3PEstimator::X_t> rig_points2D(points2D.size());
+  double error_threshold_camera = 0.;
+  for (size_t i = 0; i < points2D.size(); i++) {
+    const size_t camera_idx = camera_idxs[i];
+    rig_points2D[i].ray_in_cam =
+        cameras[camera_idx].CamFromImg(points2D[i]).homogeneous().normalized();
+    rig_points2D[i].cam_from_rig = cams_from_rig[camera_idx];
+    error_threshold_camera +=
+        cameras[camera_idx].CamFromImgThreshold(max_error_px);
+  }
+  // Average of the errors over the cameras, weighted by the number of
+  // correspondences
+  error_threshold_camera /= points2D.size();
+
+  // Associate unique ids to each 3D point.
+  // Needed for UniqueInlierSupportMeasurer to avoid counting the same
+  // 3D point multiple times due to FoV overlap in rig.
+  std::vector<Eigen::Vector3d> unique_points3D = points3D;
+  std::sort(unique_points3D.begin(), unique_points3D.end(), LowerVector3d);
+  unique_points3D.erase(
+      std::unique(unique_points3D.begin(),
+                  unique_points3D.end(),
+                  [](const Eigen::Vector3d& v1, const Eigen::Vector3d& v2) {
+                    return v1.isApprox(v2, 1e-5);
+                  }),
+      unique_points3D.end());
+  std::vector<size_t> point3D_ids;
+  point3D_ids.reserve(points3D.size());
+  for (const auto& p3D : points3D) {
```

Comment:
nit: for consistency in the repo
```suggestion
  for (const auto& point3D : points3D) {
```

Reply:
Done.